x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRepository.java

```diff
@@ -292,12 +306,47 @@ public void restoreShard(IndexShard indexShard, SnapshotId snapshotId, Version v
             store.decRef();
         }
 
+        Client remoteClient = client.getRemoteClusterClient(remoteClusterAlias);
+
         Map<String, String> ccrMetaData = indexShard.indexSettings().getIndexMetaData().getCustomData(Ccr.CCR_CUSTOM_METADATA_KEY);
         String leaderUUID = ccrMetaData.get(Ccr.CCR_CUSTOM_METADATA_LEADER_INDEX_UUID_KEY);
         Index leaderIndex = new Index(shardId.getIndexName(), leaderUUID);
         ShardId leaderShardId = new ShardId(leaderIndex, shardId.getId());
 
-        Client remoteClient = client.getRemoteClusterClient(remoteClusterAlias);
+        final String retentionLeaseId = retentionLeaseId(indexShard.shardId().getIndex().getUUID(), leaderUUID);
+        logger.trace(
+                () -> new ParameterizedMessage("{} requesting leader to add retention lease [{}]", shardId, retentionLeaseId));
+        final Optional<RetentionLeaseAlreadyExistsException> maybeAddAlready =
+                addRetentionLease(leaderShardId, retentionLeaseId, remoteClient);
+        maybeAddAlready.ifPresent(addAlready -> {
+            logger.trace(() -> new ParameterizedMessage("{} retention lease already exists, requesting a renewal", shardId), addAlready);
+            final Optional<RetentionLeaseNotFoundException> maybeRenewNotFound =
+                    renewRetentionLease(leaderShardId, retentionLeaseId, remoteClient);
+            maybeRenewNotFound.ifPresent(renewNotFound -> {
+                logger.trace(() -> new ParameterizedMessage(
+                        "{} retention lease not found while attempting to renew, requesting a final add", shardId), renewNotFound);
+                final Optional<RetentionLeaseAlreadyExistsException> maybeFallbackAddAlready =
+                        addRetentionLease(leaderShardId, retentionLeaseId, remoteClient);
+                maybeFallbackAddAlready.ifPresent(fallbackAddAlready -> {
+                    /*
+                     * At this point we tried to add the lease and the retention lease already existed. By the time we tried to renew the
+                     * lease, it expired or was removed. We tried to add the lease again and it already exists? Bail.
+                     */
+                    assert false : fallbackAddAlready;
+                    throw fallbackAddAlready;
+                });
+            });
+        });
+
+        // schedule renewals to run during the restore
+        final Scheduler.Cancellable renewable = threadPool.scheduleWithFixedDelay(
+                () -> {
+                    logger.trace("{} background renewing retention lease during restore", shardId);
+                    renewRetentionLease(leaderShardId, retentionLeaseId, remoteClient);
```

Comment:
as we except this to possibly throw an exception, should we log it here instead of bubbling it up to uncaught exception handler?

Reply:
Yes, I took care of this. There was more to do here, including authorizing for security. See a42d229cbd8fa74baa67d7074b9844b45e5120db.